HWKAPM-774 - Split OpenTracing example into individual Vert.x services

jpkrohling opened the pull request.

jpkrohling commented:
@objectiser , @pavolloffay : I don't know yet if we should keep this as a separate example, or somehow share some code between this and the `tests/app/vertx-opentracing` example. While it looks similar, I think they are too different right now. Unless we can completely replace the tests/app/vertx-opentracing, but I don't know how much effort would it be to change the tests (as it's not clear to me, yet, how it all ties together).

pavolloffay commented:
@jpkrohling can it be deployed with docker-compose?

jpkrohling commented:
Good question. Probably, yes, but the focus (so far) has been on OpenShift for this one.

pavolloffay commented:
Another question, is there in openshift something like one descriptor file which would run multiple services? (something like docker-compose)

jpkrohling commented:
> Another question, is there in openshift something like one descriptor file which would run multiple services? (something like docker-compose)

Yes. We have it on jboss-dockerfiles, deploying a Hawkular APM + Elastic Search instance, but I wanted to, on purpose, not to instruct the user to run "this single command and have it magically done". Instead, I'd tell them all the commands they have to use when deploying their own application as well.

jpkrohling commented:
I updated the PR to include a `docker-compose` on `tests/instrumentation-framework/src/test/resources/testVertxOpenTracingDockerImage`, but I'm not sure it will work. 

@pavolloffay , to which directory are the paths relative to?

objectiser commented:
@jpkrohling Good start - one main thing - could you remove the explicit setting of the 'service' tag in each of the services, so that they get implicitly set when run in OpenShift, and would also be good to set them via HAWKULAR_APM_SERVICE_NAME in the docker-compose file.

When done, let me know and I'll try running from the instructions.

jpkrohling commented:
> could you remove the explicit setting of the 'service' tag in each of the services,

@objectiser : Do you mean the `-l app=order-manager` ?

objectiser commented:
@jpkrohling Got an email notification, but can't find your question here - by explicit setting service I meant: https://github.com/hawkular/hawkular-apm/pull/706/files#diff-00a713e6c2f4f9bb612114c3ea9d2e47R78

jpkrohling commented:
> Got an email notification, but can't find your question here

I removed the question once I realized what you meant. I was probably missing a coffee or something :)

I just updated the PR with the suggestions. I also moved the example into the location of the previous test, as this might be referenced by other places, like blog posts or other readme files. As this is an evolution of that code, I think it makes sense to keep it there.

I think it's now ready for a final review.

jpkrohling commented:
retest this please

pavolloffay commented:
@objectiser Yesterday we discussed supporting also docker-compose as one deployment. What do you think? As it is used in `itest` then it wouldn't be pain to add other compose file directly to the example folder. Docker compose might be more used than OC so it wider range of users could try it. 

Another thing: Would be better to provide one configuration file for OC? It speeds up deployment and users don't have to navigate to many directories and run multiple commands. We can provide a diagram with architecture to make users aware that there are multiple containers.

jpkrohling commented:
I shared my point of view on the IRC channel already, but let me record it here as well. 

Adding support for something is easy. Maintaining it in the log term is hard (and expensive). So, I have nothing against a `docker-compose.yml` per se (I actually do have one here), but I'm afraid that it *might* get outdated in the future. Once that happens, we'll either figure this out right away, or we'll wait for an user to let us know (and disappoint every user that tried and didn't let us know that it was failing).

So, I'm in favor of having *one* target platform, in addition to bare metal. Considering that we are giving some attention to OpenShift lately, that's the platform I'd vote for.

> As it is used in itest then it wouldn't be pain to add other compose file directly to the example folder

The fact that one `docker-compose.yml` is used doesn't mean the other will remain correct.

> Docker compose might be more used than OC so it wider range of users could try it.

In general, or among our target group? If we provide instructions on how to run on bare metal (like there is), wouldn't that be sufficient for folks that don't have OpenShift?

> We can provide a diagram with architecture to make users aware that there are multiple containers.

OpenShift already has a UI with a representation of the containers/services. See, we didn't even add a `docker-compose.yml` yet and it's already becoming more expensive :-)

objectiser commented:
@jpkrohling @pavolloffay Suggestion:

* We have the example (services) under the examples folder rather than test, as it represents a demoable example.
* In the tests app folder we have a docker-compose file that takes the individual docker files for the services and enables them to be used in the instrumentation test harness - the fact that it is run on every PR means that any issues will be detected.
* In the examples folder we provide instructions for deploying each service to openshift individually
* In a blog demo'ing the example, we provide a openshift deployment config that makes it easy to deploy all services at once - but we don't necessarily ship it in code yet, until we are happy to support it, or we get requests?

What do you think?

pavolloffay commented:
@objectiser I think you did not respond to my question.

> In a blog demo'ing the example, we provide a openshift deployment config that makes it easy to deploy all services at once - but we don't necessarily ship it in code yet, until we are happy to support it, or we get requests?

Then we might change the repository code and user would not be able to run it. I think having it in repo is better. (and also easier to run the example, just clone repo and run a command)

> We can provide a diagram with architecture to make users aware that there are multiple containers.
> 
> OpenShift already has a UI with a representation of the containers/services. See, we didn't even add a docker-compose.yml yet and it's already becoming more expensive :-)

This applies if we provide docker-compose support. 

I still think that supporting docker-compose would be good for us as it is widely used and it is **going to be used in tests**. It's good for us to target a broader audience. 

Docker compose supports merging compose files [1]. We could provide compose.yaml in example folder and add `itest` specific properties in compose.yaml in test folder. (this would require simple change in instrumentation-framework)

[1]: https://docs.docker.com/compose/extends/

objectiser commented:
@pavolloffay Did not mean move out of the repo - but under the examples folder.

Why would we need more than one docker-compose.yml?

objectiser commented:
@pavolloffay You mean the question about single OC? If so, I covered it in my response - to deal with Juca's concern, I thought we could put the OC deployment config in the blog so users can use it to quickly get up and running. But outside codebase so we don't have to worry about keeping it maintained.

pavolloffay commented on tests/app/vertx-opentracing/README.md:
````diff
@@ -12,44 +12,63 @@ Manager is performed via publishing a message on a shared destination.
-```shell
-$ docker run -it --rm -p 8180:8180 -e HAWKULAR_APM_URI=http://172.17.0.1:8080 -e HAWKULAR_APM_USERNAME=jdoe -e HAWKULAR_APM_PASSWORD=password hawkular/apm-tests-app-vertx-opentracing
-```
+    sudo iptables -F
````
It would be good to put here `mvn install` otherwise images builds will fail

pavolloffay commented on tests/app/vertx-opentracing/README.md:
```diff
@@ -12,44 +12,63 @@ Manager is performed via publishing a message on a shared destination.
+
+Once all microservices are set and ready, the following command can be issued to test:
+
+    curl -X POST -d '{"accountId":"fred","itemId":"laptop","quantity":2}' http://order-manager-myproject.192.168.178.26.xip.io/orders
```
Could you change the IP to something like `<oc cluster IP>`?

pavolloffay commented on tests/app/vertx-opentracing/README.md:
```diff
@@ -12,44 +12,63 @@ Manager is performed via publishing a message on a shared destination.
+Once all microservices are set and ready, the following command can be issued to test:
+
+    curl -X POST -d '{"accountId":"fred","itemId":"laptop","quantity":2}' http://order-manager-myproject.192.168.178.26.xip.io/orders
+
```
Maybe it would be good to add go to Hawkular APM UI `https://hawkular-apm-myproject.<oc cluser IP>.xip.io`

pavolloffay replied:
And also what is the user:password or where they can find it

pavolloffay commented:
@objectiser 
> Why would we need more than one docker-compose.yml?
One for running example and other for `itest`. docker-compose.yaml for `itest` would inherit from one from example folder and add `itest` specific config.

> A blog post can always be updated if someone was to go back to it and things had changed in the meantime. If it is in the repo, then we would need to find a way to hook it into integration tests to make sure it didn't get broken - or have a manual test procedure before release, which is not ideal.

Manual maintaining needs also polyglot-zipkin. You would quicker find issues with multiple container environment if it can be run with oneliner rather than multiple command spread in multiple dirs. (it's just my opinion)

objectiser commented:
@pavolloffay Understand your perspective, but @jpkrohling point is also valid.

With polyglot-zipkin, it has a docker-compose.yml so we could hook it into the instrumentation test framework to automate it.

@theute @pilhuhn Could you help us resolve a dilemma - we have an example (opentracing-vertx) that we will use for demos on OpenShift, but is also used as an integration test. For testing, we have a docker-compose.yml that is automatically invoked and therefore tested with each PR.

The issue is whether we should include an OpenShift deployment config, to help deploy all services in one step, or just provide instructions for individual service install. Juca's concern is that this single deployment config may become out of date (hard to maintain) as we don't have any way to automate testing it. MSA helloworld example only provides instructions for individual service deploy.

So options are: (1) we provide a blog that shows a single step install of all services, but in the instructions in the repo we just deploy as individual services. Or (2) we include the single install deployment config in the example folder. Thoughts?

pavolloffay commented:
If we have one config file to run it all in one, then it is also easier for us to maintain/test/demo it. I think demos should be easy to run, otherwise users would not want to try it.

pilhuhn commented:
I am in favour of 2) the easier it is for the user to get going the better. A blog/docs can still explain details. Getting the example to quickly run is important.

theute commented:
Not sure to understand al (in particular "all" services)l, so rather than trying to reply 1) or 2) I would say that my preference is that:
   - I can install APM server on OpenShift with the template I've seen already. (just the server + ES)
   - I can then install some examples on OpenShift in a similar way (Ideally 1 command to execute per example)
And we need to point to files that will be maintained.

objectiser commented:
@theute @pilhuhn Thanks for your input.

@jpkrohling Sorry, but looks like the main opinion is that it is better to have a single OpenShift deploy command - I guess an 'instant app' rather than template.

jpkrohling commented:
Alright, I'll add a YAML file for that :) 

So, in short, the following is necessary for this PR:

- [ ] Review changes requested by Pavol
- [ ] Move from `tests/app/` to `examples/`
- [ ] Add a `deployment.yml`, allowing users to run `oc create -f deployment.yml` and have all services deployed
- [ ] Blog post about this example

Is that correct?

objectiser commented:
@jpkrohling Yes I think that is it - just to check, the instrumentation test framework will operate on that example?

pavolloffay commented:
To be clear did not request to move it to `examples` folder. If this code will be used by `itest` then it's better to keep it in `tests/app`. What do you think?

jpkrohling commented:
@objectiser wrote:
> just to check, the instrumentation test framework will operate on that example?

Yes, no matter where the example is located, the instrumentation tests will use it.

@pavolloffay wrote:
> If this code will be used by `itest` then it's better to keep it in `tests/app`. What do you think?

Good question. Having it on `examples` makes it easier for users to find, but `tests/app` makes it easier for developers to maintain.

objectiser commented:
If it is going to be an example that we would like OpenShift users to find and use, then my preference would be to have it under examples and used by the `itests`. We could still have the tests/app/vertx-opentracing folder for reference to the example - both in case of existing links, but also as a placeholder to make clear that it is used as part of the integration tests?

jpkrohling commented on tests/instrumentation-framework/src/test/resources/testOpenTracingPolyglot/test.json:
```diff
@@ -9,7 +9,7 @@
     {
       "description" : "Context propagation",
-      "skip": false,
+      "skip": true,
```
And I was thinking about how lucky I am that I got two build successes in a row!

jpkrohling commented on tests/instrumentation-framework/src/test/resources/testVertxOpenTracingDockerImage/docker-compose.yml:
```diff
@@ -0,0 +1,61 @@
+#
```
Ok, makes sense.

jpkrohling commented on tests/app/vertx-opentracing/account-manager/src/main/resources/vertx-default-jul-logging.properties:
```diff
@@ -0,0 +1,28 @@
+# limitations under the License.
+#
+
+handlers=java.util.logging.ConsoleHandler,java.util.logging.FileHandler
```
For a production-quality application, this could indeed be a good idea. But for this example, I'd rather leave it there. I tried to quickly put things in `commons`, and I got the default logging from JUL.

jpkrohling commented:
PR updated with the changes requested by Pavol during the review. The other changes are on the queue.

jpkrohling commented:
PR updated, to include a "single command" setup for the examples. On my machine, the command takes 3m to finish.

Also, we have already some resources pointing to the example at `tests/app`. Should we really move it to `/examples` ?

objectiser commented:
@jpkrohling We should updated the ones we know about, but leave a folder in the old location with a README.md referring to the new location?

jpkrohling commented:
That's an option, yes.

objectiser commented:
@jpkrohling Ansible script works great now. Will just add a few suggestions for the readme.

objectiser commented on examples/vertx-opentracing/README.md:
```diff
@@ -0,0 +1,153 @@
+
+
+![Distributed tracing - Instance view](screenshot-2.png?raw=true "Distributed tracing - Instance view")
+
```
It might be good to have a heading here, something like "Running on a local OpenShift cluster" - and then have the two options under that heading, with the "Running outside OpenShift .." at the same level as this new header.

objectiser replied:
In this new section, we may also need to provide basic instructions to install openshift?

objectiser commented on examples/vertx-opentracing/README.md:
```diff
@@ -0,0 +1,153 @@
+![Distributed tracing - Instance view](screenshot-2.png?raw=true "Distributed tracing - Instance view")
+
+## Setting it up - Option 1 - Automated with Ansible
+
```
Need to indicate dependency on ansible, and the install command?

objectiser commented on examples/vertx-opentracing/README.md:
```diff
@@ -0,0 +1,153 @@
+which can be seen on the OpenShift UI:
+
+    curl -X POST -d '{"accountId":"fred","itemId":"laptop","quantity":2}' http://order-manager-myproject.<your OpenShift IP>.xip.io/orders
+
```
Same as above - link to section describing further requests.

objectiser replied:
Wondering whether it needs the `oc get route order-manager` text here as well? Ideally it would be good to have it in one place - may be the text for trying on openshift could be in one place - but then the user would need to be directed from the two places?

objectiser commented on examples/vertx-opentracing/README.md:
```diff
@@ -0,0 +1,153 @@
+To test, run:
+
+    curl -X POST -d '{"accountId":"fred","itemId":"laptop","quantity":2}' http://127.0.0.1:8180/orders
+
```
'See following section for futher requests' - maybe redundant but consistent with above?

objectiser commented on examples/vertx-opentracing/README.md:
```diff
@@ -0,0 +1,153 @@
+Then, each microservice has to be started, possibly on its own terminal (don't forget to set the env vars above on each terminal).
+From the directory of this README file, run:
+
+    java -jar account-manager/target/hawkular-apm-tests-app-vertx-opentracing-account-manager-fat.jar -cluster
```
Could you also add `-DHAWKULAR_APM_SERVICE_NAME=....` to each command, to show setting the service externally?

objectiser commented on examples/vertx-opentracing/README.md:
```diff
@@ -135,7 +135,7 @@ See [Example Requests](#example-requests) for further commands.
 
-For the following examples, we'll use `${ORDER_MANAGER_SERVER}` to refer to the Hawkular URI. It might be a simple hostname,
+For the following examples, we'll use `${ORDER_MANAGER_SERVER}` to refer to the example's server. It might be a simple hostname,
 like when it's deployed on OpenShift, or `hostname:port`, like when it's deployed locally (`localhost:8080`, for instance).
```
Only one change - if you could make it `8180` as this is the port outside openshift.

objectiser commented:
@jpkrohling Great job! Only the one remaining comment about the port to use, and after that fine to squash/merge.

jpkrohling commented:
Done. Before I squash, it would be nice to get someone "fresh" to test it. 

@pilhuhn : would you give it a try ?

pavolloffay commented:
I'm going to try it too :+1:

pavolloffay commented on examples/vertx-opentracing/vertx-opentracing.yml:
```diff
@@ -0,0 +1,85 @@
+      shell: "oc get route order-manager | grep order-manager-order | awk '{print $2}'"
+      register: route_order_manager
+
+    - name: Print out the order-manager address
```
Could you please print address of oc cluster, Hawkular APM server and order manager

objectiser commented:
@jpkrohling Thanks, good job!
